File: firmware_factory_control_unit/builder/src/board-context.ts
// "Letztes bekanntes Board"-Caching (s. docs/build-process.md Abschnitt 4): Phasen, die kein
// eigenes Board auslesen (Register-Map, Git-Status, Copy, Flash), muessen trotzdem wissen, in
// welchen Board-Archiv-Ordner sie schreiben/aus welchem sie lesen sollen. Der zuletzt von
// ReadHardwareIDsAndGenerateFilesAndCerts{Lazy,Forced} erkannte boardId wird dafuer in einer
// gitignoreten Datei unter build/ gecacht.
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { boardIdCacheFile, buildDir } from "./paths.ts";

export function readCachedBoardId(): string | undefined {
	if (!existsSync(boardIdCacheFile)) return undefined;
	const content = readFileSync(boardIdCacheFile, "utf8").trim();
	return content.length > 0 ? content : undefined;
}

export function writeCachedBoardId(boardId: string): void {
	mkdirSync(buildDir, { recursive: true });
	writeFileSync(boardIdCacheFile, boardId);
}

export function resolveBoardId(explicit?: string): string | undefined {
	return explicit ?? readCachedBoardId();
}

// Fuer Phasen, die ohne Board-Kontext keinen Sinn ergeben (Copy, Flash) -- klare Fehlermeldung
// statt eines kryptischen ENOENT beim Zugriff auf einen nicht existierenden Archiv-Ordner.
export function requireBoardId(explicit?: string): string {
	const boardId = resolveBoardId(explicit);
	if (!boardId) {
		throw new Error(
			"Kein Board-Kontext bekannt (weder --board angegeben noch build/.last-board-id vorhanden). " +
				"Zuerst 'ReadHardwareIDsAndGenerateFilesAndCertsLazy' mit angeschlossenem Board ausfuehren."
		);
	}
	return boardId;
}

File: firmware_factory_control_unit/builder/BoardContext.cs
// "Letztes bekanntes Board"-Caching (s. docs/build-process.md Abschnitt 4).
namespace Builder;

public static class BoardContext
{
	public static string? ReadCachedBoardId()
	{
		if (!File.Exists(Paths.BoardIdCacheFile)) return null;
		var content = File.ReadAllText(Paths.BoardIdCacheFile).Trim();
		return content.Length > 0 ? content : null;
	}

	public static void WriteCachedBoardId(string boardId)
	{
		Directory.CreateDirectory(Paths.BuildDir);
		File.WriteAllText(Paths.BoardIdCacheFile, boardId);
	}

	public static string? ResolveBoardId(string? explicitBoardId) => explicitBoardId ?? ReadCachedBoardId();

	// Fuer Phasen, die ohne Board-Kontext keinen Sinn ergeben (Copy, Flash) -- klare Fehlermeldung
	// statt eines kryptischen Fehlers beim Zugriff auf einen nicht existierenden Archiv-Ordner.
	public static string RequireBoardId(string? explicitBoardId)
	{
		var boardId = ResolveBoardId(explicitBoardId);
		if (boardId is null)
		{
			throw new InvalidOperationException(
				"Kein Board-Kontext bekannt (weder --board angegeben noch build/.last-board-id vorhanden). " +
				"Zuerst 'ReadHardwareIDsAndGenerateFilesAndCertsLazy' mit angeschlossenem Board ausfuehren.");
		}
		return boardId;
	}
}
